tests/functional/autoprompt/test_prompttoolkit.py

```diff
@@ -146,3 +234,273 @@ def test_handle_args_with_spaces(self):
         self.assertEqual(prompter.input_buffer.document.text,
                          "iam create-role --description 'With spaces' ")
         self.assertEqual(args, original_args)
+
+
+class TestPromptToolkitDocBuffer(BasicPromptToolkitTest):
+    def get_updated_doc_buffer_text(self, original_args):
+        self.prompter.args = original_args
+        self.prompter.pre_run()
+        return self.prompter.doc_buffer.document.text
+
+    def test_doc_buffer_not_shown_on_start_and_not_focusable(self):
+        stubber = ApplicationStubber(self.prompter.create_application())
+        stubber.add_key_assert(None, lambda app: app.show_doc is False)
+        stubber.add_key_assert(
+            Keys.F1, lambda app: app.current_buffer.name == 'input_buffer'
+        )
+        stubber.run(self.prompter.pre_run)
+
+    def test_doc_buffer_shows_hides_on_F3(self):
+        stubber = ApplicationStubber(self.prompter.create_application())
+        stubber.add_key_assert(Keys.F3, lambda app: app.show_doc is True)
+        stubber.add_key_assert(
+            Keys.F3, lambda app: app.current_buffer.name == 'input_buffer'
+        )
+        stubber.run(self.prompter.pre_run)
+
+    def test_doc_buffer_gets_and_removes_focus(self):
+        stubber = ApplicationStubber(self.prompter.create_application())
+        stubber.add_key_assert(Keys.F3, lambda app: app.show_doc is True)
+        stubber.add_key_assert(
+            Keys.F2,
+            lambda app: app.current_buffer.name == 'doc_buffer'
+        )
+        stubber.add_key_assert(
+            'q',
+            lambda app: app.current_buffer.name == 'input_buffer'
+        )
+        stubber.run(self.prompter.pre_run)
+
+    def test_doc_buffer_keeps_position_if_content_dont_change(self):
+        original_args = ['ec2', 'describe-instances']
+        self.prompter.args = original_args
+        stubber = ApplicationStubber(self.prompter.create_application())
+        stubber.add_key_assert(Keys.F3)
+        stubber.add_key_assert(Keys.F2)
+        # go to the top row
+        stubber.add_key_assert('g', lambda app: app.layout.get_buffer_by_name(
+                    'doc_buffer').document.cursor_position_row == 0)
+        # go three rows down
+        stubber.add_key_assert('j')
+        stubber.add_key_assert('j')
+        stubber.add_key_assert('j', lambda app: app.layout.get_buffer_by_name(
+                    'doc_buffer').document.cursor_position_row == 3)
+        # get back to input
+        stubber.add_key_assert('q')
+        # go on enter the rest of the command and  check that cursor in doc
+        # panel remains on the same place
+        for k in ' --instances':
+            stubber.add_key_assert(k, lambda app: app.layout.get_buffer_by_name(
```

Comment:
Is it possible to add assertions to the body for these? For example, we'd be able to define this as:
```python
lambda app: assert app.layout.get_buffer_by_name('doc_buffer').document.cursor_position_row == 3)
```
This would be nice because we would be able to make writing assertions a bit more intuitive and more powerful (e.g. have multiple assertions in a single function, asserting any caught errors, etc.)

Reply:
Yeah, it would be cleaner but unfortunately it's not a valid Python code